controllers/global_search_setup.go

```diff
@@ -0,0 +1,570 @@
+				Message:            "Failed to enable global search. " + err.Error(),
+				LastTransitionTime: metav1.Now(),
+			})
+			if updateErr != nil {
+				log.Error(updateErr, "Failed to update Global Search status condition on Search CR instance.")
+			}
+		} else {
+			updateErr := r.updateGlobalSearchStatus(ctx, instance, metav1.Condition{
+				Type:               "GlobalSearchReady",
+				Status:             metav1.ConditionTrue,
+				Reason:             "None",
+				Message:            "None",
+				LastTransitionTime: metav1.Now(),
+			})
+			if updateErr != nil {
+				log.Error(updateErr, "Failed to update the Global Search status condition on Search CR instance.")
+			}
+		}
+	} else {
+		log.Info("The global-search-preview annotation is not present. Checking if global search was enabled before.")
+
+		// Use the status conditions to determine if global search was enabled before this reconcile.
+		globalSearchPresent := false
+		globalSearchConditionIndex := -1
+		for condIndex, condition := range instance.Status.Conditions {
+			if condition.Type == "GlobalSearchReady" {
+				globalSearchPresent = true
+				globalSearchConditionIndex = condIndex
+				break
+			}
+		}
+		if globalSearchPresent {
+			err := r.disableGlobalSearch(ctx, instance)
+			if err != nil {
+				log.Error(err, "Failed to disable global search.")
+				updateErr := r.updateGlobalSearchStatus(ctx, instance, metav1.Condition{
+					Type:               "GlobalSearchReady",
+					Status:             metav1.ConditionFalse,
+					Reason:             "GlobalSearchCleanupFailed",
+					Message:            "Failed to disable global search. " + err.Error(),
+					LastTransitionTime: metav1.Now(),
+				})
+				if updateErr != nil {
+					log.Error(updateErr, "Failed to update Search CR instance status.")
+				}
+				return &reconcile.Result{}, err
+			}
+
+			// Remove the status condition.
+			instance.Status.Conditions = append(instance.Status.Conditions[:globalSearchConditionIndex],
+				instance.Status.Conditions[globalSearchConditionIndex+1:]...)
+
+			err = r.commitSearchCRInstanceState(ctx, instance)
+			if err != nil {
+				log.Error(err, "Failed to update Search CR instance status.")
+				return &reconcile.Result{}, err
+			}
+		}
+	}
+	return &reconcile.Result{}, nil
+}
+
+// Validate pre-requisites.
+//
+//	a. MulticlusterGlobalHub operator is installed in the cluster.
+//	b. The ManagedServiceAccount add-on is enabled in the MultiClusterEngine CR.
+//	c. The ClusterProxy addon is enabled in the MultiClusterEngine CR.
+func (r *SearchReconciler) validateGlobalSearchPrerequisites(ctx context.Context) error {
+	log.V(5).Info("Checking global search pre-requisites.")
+	// Verify that MulticlusterGlobalHub operator is installed.
+	// oc get multiclusterglobalhub -A
+	multiclusterGlobalHub, err := r.DynamicClient.Resource(multiclusterGlobalHubGvr).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("MulticlusterGlobalHub operator is not installed.")
```

Comment:
There can be other errors other than 'Not Found' if the client failed to list. Should we attach the actual error too?

Reply:
good point. I wanted to keep it simple because the returned error gets attached to the condition message. What do you think about making it more generic like "Unable to validate the prerequisite MulticlusterGlobalHub operator."
I'll add a log with the full error.

Reply:
Failed to verify if MulticlusterGlobalHub operator is installed, maybe?

Reply:
"Unable to validate the prerequisite MulticlusterGlobalHub operator" sounds better. :)